doc/usage.rst

```diff
@@ -187,6 +187,76 @@ names of these files in the registry **must use Unix-style separators** (``'/'``
 on Windows. We will handle the appropriate conversions.
 
 
+Post-processing hooks
+---------------------
+
+Sometimes further post-processing actions need to be taken on downloaded files
+(unzipping, conversion to a more efficient format, etc). If these actions are time or
+memory consuming, it would be best to do this only once when the file is actually
+downloaded and not every time :meth:`pooch.Pooch.fetch` is called.
+
+One way to do this is using *post-processing hooks*. Method :meth:`pooch.Pooch.fetch`
+takes a ``hook`` argument that allows us to specify a function that is executed
+post-download and before returning the local file path. The hook also lets us overwrite
+the file name returned by :meth:`pooch.Pooch.fetch`.
+
+For example, let's say our data file is zipped and we want to store an unzipped copy of
+it and read that instead. We can do this with a post-processing hook that unzips the
+file and returns the path to the unzipped file instead of the original zip archive:
+
+.. code:: python
+
+    import os
+    from zipfile import ZipFile
+
+    def unpack_hook(fname, action, pup):
+        """
+        Post-processing hook to unzip a file and return the unzipped file name.
+
+        Parameters
+        ----------
+        fname : str
+           Full path of the zipped file in local storage
+        action : str
+           One of "download" (file doesn't exist and will download),
+           "update" (file is outdated and will download), and
+           "fetch" (file exists and is updated so no download).
+        pup : Pooch
+           The instance of Pooch that called the hook function.
+
+        Returns
+        -------
+        fname : str
+           The full path to the unzipped file.
+           (Return the same fname is your hook doesn't modify the file).
+
+        """
+        # Create a new name for the unzipped file
+        unzipped = fname + ".unzipped"
+        # Don't unzip if file already exists and is not being downloaded
+        if action in ("update", "download") or not os.path.exists(unzipped):
+            with ZipFile(fname, "r") as zip_file:
+                # Extract the data file from within the archive
```

Comment:
Isn't better to use `ZipFile.extract` rather than `ZipFile.open`? For example, if the desired zipped files is inside a directory (within the zip file), I think the `open` method would raise an error due to missing unzipped directory, while the `extract` would create it.

Reply:
I was using it at first but the problem is that you can't set an output file name in `extract`, just the directory. So for this example case, this is the best approach.

> if the desired zipped files is inside a directory (within the zip file)

It shouldn't be a problem because we're not saving with that file name, we're using `unzipped` instead.

An alternative way if you have multiple files would be to set `unzipped` as the output directory and then use `extract`. We shouldn't extract into the data directory because there might be file name conflicts that are out of our control.

Reply:
:+1: